core/src/main/scala/org/broadinstitute/dsde/rawls/billing/BpmBillingProjectCreator.scala

```diff
@@ -0,0 +1,57 @@
+package org.broadinstitute.dsde.rawls.billing
+
+import akka.http.scaladsl.model.StatusCodes
+import org.broadinstitute.dsde.rawls.RawlsExceptionWithErrorReport
+import org.broadinstitute.dsde.rawls.model.{CreateRawlsV2BillingProjectFullRequest, ErrorReport, UserInfo}
+
+import scala.concurrent.{ExecutionContext, Future, blocking}
+
+
+/**
+ * This class knows how to validate Rawls billing project requests and instantiate linked billing profiles in the
+ * billing profile manager service.
+ */
+class BpmBillingProjectCreator(billingRepository: BillingRepository,
+                               billingProfileManagerDAO: BillingProfileManagerDAO)
+                              (implicit val executionContext: ExecutionContext) extends BillingProjectCreator {
+
+  override def validateBillingProjectCreationRequest(createProjectRequest: CreateRawlsV2BillingProjectFullRequest,
```

Comment:
From my Java memories, I would expect this method to either return a boolean (indicating whether it is valid), or throw a checked exception. I see that this is using exceptions to indicate if something is invalid. In Scala are you not encouraged to declare checked exceptions that callers should be prepared to handle (vs. undeclared runtime errors that are indication of code errors)?

Reply:
I don't think it's _automatically_ an indication of a code error if an exception propagates up. It kind of just depends on how your service is structured. 
In this service, I think RawlsExceptionWithErrorReport is intended to be an exception that is caught higher up, and not necessarily handled by the caller, in order to propagate to the user. 

You raise a good point about NotImplementedError though. 
IMO, it's probably fine if you don't expect to hit it (making it more of a sanity check sort of thing), but if it's more validation that might be too low level to throw as-is.

ManagedAppNotFoundException extends RawlsExceptionWithErrorReport so as far as I can tell, should be fine to throw bringing explicit attention to it.

Reply:
Scala does not have checked exceptions, they are all runtime (at least in scala 2, I am unsure about 3). 

Re: `NotImplementedError` I want some kind of precondition here as a sanity check as @blakery mentions; I am fine raising some other exception, but the intention is for this to be a "this should never happen" assertion.

Reply:
I can see not mentioning `NotImplementedError`, but I think it should be documented in some way how the method indicates that the request isn't valid. Ideally I would expect that to be with "throws ManagedAppNotFoundException", but since that isn't possible, documentation seems like it would be helpful. Another alternative would be to follow this pattern:

```
More generally, Either[E,V] has been used as a return type with the convention that returning Left[E] holds an error E, such as an exception, while Right[V] returns a successful value of type V. Try[V] works similarly.
```

It looks like checked exceptions are being considered for Scala (which is where I also got ^)-- https://medium.com/scala-3/safer-exceptions-in-scala-3-12c47100f050.

Reply:
I like the idea of returning a semantically clearer `Either[E, V]`. That being said, most of the consuming code for this class is expecting a failed future rather than a successful future. 

Also given the general approach in Rawls seems to be to throw and fail the future, I am inclined to stick with the current approach for consistency and update the docs.

Reply:
Sure, that's fine… I've come around to it now.